File: lib/providers/youtube.ts
import { prisma } from '../prisma';
import { UploadTarget, Video } from '@prisma/client';
import { OAuth2Client } from 'google-auth-library';
import { google } from 'googleapis';
import { getReadUrl } from '../storage';

function getOAuthClient(account: any) {
  const client = new OAuth2Client(process.env.GOOGLE_CLIENT_ID!, process.env.GOOGLE_CLIENT_SECRET!);
  client.setCredentials({
    access_token: account.accessToken,
    refresh_token: account.refreshToken,
    expiry_date: account.tokenExpiresAt ? new Date(account.tokenExpiresAt).getTime() : undefined,
  });
  return client;
}

export async function handleYouTubeUpload(target: UploadTarget & { video: Video; account: any }) {
  const oAuth2Client = getOAuthClient(target.account);
  const youtube = google.youtube({ version: 'v3', auth: oAuth2Client });

  // Signed GET URL to fetch the video from S3/R2
  const readUrl = await getReadUrl(target.video.s3Key);

  // Prepare metadata
  const privacyStatus = target.video.visibility === 'PRIVATE' ? 'private' : target.video.visibility === 'SCHEDULED' ? 'private' : 'public';

  // Create a readable stream from the remote file (Node 18+ has global fetch)
  const res = await fetch(readUrl);
  if (!res.ok || !res.body) throw new Error(`Failed to read source video: HTTP ${res.status}`);

  await prisma.uploadTarget.update({ where: { id: target.id }, data: { progress: 20 } });

  // Upload via googleapis client (resumable handled internally)
  const insertRes = await youtube.videos.insert({
    part: ['snippet', 'status'],
    requestBody: {
      snippet: {
        title: target.video.title,
        description: target.video.description ?? '',
        tags: target.video.tags,
      },
      status: {
        privacyStatus,
        publishAt: target.video.scheduledAt ?? undefined,
        selfDeclaredMadeForKids: false,
      },
    },
    media: {
      body: res.body as any,
    },
  } as any);

  const videoId = insertRes.data.id;
  if (!videoId) throw new Error('YouTube upload did not return a video ID');

  await prisma.uploadTarget.update({ where: { id: target.id }, data: { progress: 90 } });

  // Optionally set custom thumbnail later if available (thumbnails.set requires multipart upload)
  const remoteUrl = `https://www.youtube.com/watch?v=${videoId}`;
  await prisma.uploadTarget.update({ where: { id: target.id }, data: { remoteId: videoId, remoteUrl } });
}

File: backend/youtube.go
package main

import (
	"context"
	"log"
	"os"
	"time"

	"cloud.google.com/go/firestore"
	"google.golang.org/api/youtube/v3"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

// uploadVideoToYouTube uploads the video from GCS to the channel linked to the account.
// It updates the target in Firestore with remoteId, remoteUrl on success, or status ERROR on failure.
func (s *Server) uploadVideoToYouTube(ctx context.Context, video Video, target UploadTarget, account Account) error {
	clientID := os.Getenv("GOOGLE_CLIENT_ID")
	clientSecret := os.Getenv("GOOGLE_CLIENT_SECRET")
	bucketName := os.Getenv("GCS_BUCKET")
	if clientID == "" || clientSecret == "" {
		return errConfig("GOOGLE_CLIENT_ID / GOOGLE_CLIENT_SECRET manquants")
	}
	if bucketName == "" {
		return errConfig("GCS_BUCKET manquant")
	}

	// OAuth2 client with account tokens (refresh if expired)
	conf := &oauth2.Config{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		Endpoint:     google.Endpoint,
		Scopes:       []string{"https://www.googleapis.com/auth/youtube.upload", "https://www.googleapis.com/auth/youtube"},
	}
	var expiry time.Time
	if account.TokenExpiresAt != nil {
		expiry = *account.TokenExpiresAt
	}
	tok := &oauth2.Token{
		AccessToken:  account.AccessToken,
		RefreshToken: account.RefreshToken,
		Expiry:       expiry,
	}
	tokenSource := oauth2.ReuseTokenSource(tok, conf.TokenSource(ctx, tok))
	client := oauth2.NewClient(ctx, tokenSource)

	// YouTube service
	yt, err := youtube.New(client)
	if err != nil {
		return err
	}

	// GCS reader for the video file
	if s.gcsClient == nil {
		return errConfig("stockage GCS non configuré")
	}
	bucket := s.gcsClient.Bucket(bucketName)
	obj := bucket.Object(video.S3Key)
	reader, err := obj.NewReader(ctx)
	if err != nil {
		return err
	}
	defer reader.Close()

	// Privacy status
	privacy := "public"
	switch video.Visibility {
	case "PRIVATE":
		privacy = "private"
	case "SCHEDULED":
		privacy = "private" // YouTube: schedule with publishAt
	case "UNLISTED":
		privacy = "unlisted"
	}

	req := &youtube.Video{
		Snippet: &youtube.VideoSnippet{
			Title:       video.Title,
			Description: video.Description,
			Tags:        video.Tags,
		},
		Status: &youtube.VideoStatus{
			PrivacyStatus: privacy,
			SelfDeclaredMadeForKids: false,
		},
	}
	if video.Visibility == "SCHEDULED" && video.ScheduledAt != nil {
		req.Status.PublishAt = video.ScheduledAt.Format(time.RFC3339)
	}

	call := yt.Videos.Insert([]string{"snippet", "status"}, req).Context(ctx)
	res, err := call.Media(reader).Do()
	if err != nil {
		return err
	}
	if res.Id == "" {
		return errConfig("YouTube n'a pas renvoyé d'id vidéo")
	}

	remoteURL := "https://www.youtube.com/watch?v=" + res.Id
	tRef := s.fs.Collection("uploadTargets").Doc(target.ID)
	_, err = tRef.Update(ctx, []firestore.Update{
		{Path: "status", Value: "PUBLISHED"},
		{Path: "progress", Value: 100},
		{Path: "remoteId", Value: res.Id},
		{Path: "remoteUrl", Value: remoteURL},
	})
	if err != nil {
		log.Printf("failed to update target %s: %v", target.ID, err)
	}
	return nil
}

type errConfig string

func (e errConfig) Error() string { return string(e) }
